src/myrtlespeech/run/run.py

```diff
@@ -170,17 +176,123 @@ def on_train_end(self, **kwargs):
 
 
 class Saver(ModelCallback):
-    """TODO"""
+    """Loads and saves the ``model``\'s ``state_dict``.
+
+    This callback saves the ``model``'s ``state_dict`` to ``log_dir``
+    ``on_epoch_end`` with file name ``state_dict_<EPOCH>.pt``, where
+    ``<EPOCH>`` is the number of completed epochs.
+
+    At the start of training, ``on_train_begin``, the callback attempts to load
+    the model. If ``load_fp is not None`` it attempts to restore the model from
+    the ``state_dict`` saved in ``load_fp``. If ``load_fp is None`` it finds
+    the file name in ``log_dir`` with highest ``<EPOCH>`` and attempts to
+    restore the model from the ``state_dict`` saved in it.
+
+    Args:
+        log_dir: A pathlike object giving the logging directory.
+
+        load_fp: An Optional pathlike object specifying the filepath of the
+            state_dict.
+
+        model: A :py:class:`torch.nn.Module` to be saved each epoch. To enable
+            resumption of training after a crash, the user should initialise
+            this callback with a :py:class:`SeqToSeq` instance.
+    """
 
-    def __init__(self, log_dir: Union[str, Path], *args, **kwargs):
-        self.log_dir = log_dir
+    def __init__(
+        self,
+        log_dir: Union[str, Path],
+        load_fp: Optional[Union[str, Path]] = None,
+        *args,
+        **kwargs,
+    ):
+        self.log_dir = Path(log_dir)
+        self.load_fp = load_fp
         super().__init__(*args, **kwargs)
 
+    def on_train_begin(self, **kwargs) -> Optional[Dict]:
+        if self.load_fp is not None:
+            epoch, total_train_batches = self._load_seq_to_seq(self.load_fp)
+        else:
+            epoch, total_train_batches = self._load_most_recent_state_dict()
+
+        # Update CallbackHandler state_dict
+        if epoch is not None:
+            kwargs["epoch"] = epoch + 1
+        if total_train_batches is not None:
+            kwargs["total_train_batches"] = total_train_batches
+
+        if kwargs["epoch"] > kwargs["epochs"]:
+            warnings.warn(
+                f'cb_handler.state_dict["epoch"] is > '
+                f'cb_handler.state_dict["epochs"] so no training will occur.'
+            )
+        return kwargs
+
+    def _load_most_recent_state_dict(
+        self,
+    ) -> Tuple[Optional[int], Optional[int]]:
+        """Loads most recent state_dict in ``self.log_dir``.
+
+        ''Most recent'' in this case is chosen according to state filenames
+        which must be in the form 'state_dict_<EPOCH>.pt'.
+
+        Returns:
+            A Tuple of Optional ints ``epoch, total_train_batches``.
+        """
+        fnames = []
+        for fname in os.listdir(self.log_dir):
+            match = re.findall(r"state_dict_(\d+)\.pt", fname)
```

Comment:
I think I actually want the `findall` version since this specifically returns the epoch number (i.e. it matches the whole regex but returns the matches in parentheses). 

with this I can do:
```python
if match:
    fnames.append((int(match[0]), fname))
fnames.sort()
```
And I later check that the epoch in the filename is the same as the epoch in the state dict and raise a warning if not.